doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,245 @@
+| 0x1D | connection_close 0x1D | | | &#10004; | &#10004; | &#10004; | | | |
+| 0x1E | handshake_done | | | | &#10004; | | | | |
+
+The various fields are as defined in RFC 9000, although without the visual
+separation between _Pkts_ and _Spec_.
+
+| Field | Description | Type |
+| :---: | --- | --- |
+| I | Valid in Initial packets | _Pkts_ |
+| H | Valid in Handshake packets | _Pkts_ |
+| 0 | Valid in 0-RTT packets | _Pkts_ |
+| 1 | Valid in 1-RTT packets | _Pkts_ |
+| N | Packets containing only frames with this marking are not ack-eliciting. | _Spec_ |
+| C | Packets containing only frames with this marking do not count toward bytes in flight for congestion control purposes. | _Spec_ |
+| P | Packets containing only frames with this marking can be used to probe new network paths during connection migration. | _Spec_ |
+| F | The contents of frames with this marking are flow controlled. | _Spec_ |
+
+### Notes
+
+- It is likely that a wrapper structure around the individial frames will be required.
+
+### Packets
+
+Frames are coalesced into packets which are then sent by the record layer.
+The `packet_header` is a pointer to the leading bytes of the packet.
+The `frames` are pointers to the individual frames that make up the packet's body.
+It is expected that the record layer will encrypt from the `packet_header` and
+`frames` directly without a copy.
+
+```c
+struct ossl_quic_packet_st {
+    SSL *connection;
+    unsigned char *packet_header;
+    size_t packet_header_length;
+    STACK_OF(OSSL_QUIC_FRAME) *frames;
+
+    QUIC_PN packet_number; /* RFC 9000 12.3 */
+    size_t packet_length;
+
+    /* Pkts - might be better as one field as per ACK Manager */
+    unsigned int initial : 1;
+    unsigned int handshake : 1;
+    unsigned int rtt0 : 1;
+    unsigned int rtt1 : 1;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+#### Notes
+
+- `QUIC_PN` defined as uint64_t in ACK Manager.
+- The `QUIC_ACKM_TX_PKT` struture from the ACK Manager overlaps this significantly and the two could be merged.
+- Do we need the distinction between 0-rtt and 1-rtt when both are in the Application Data number space?
+- the `packet_header` will become a structure / union most likely.
+
+## Interactions
+
+The packetiser needs to interact with other modules.  This defines the APIs
+by which it does so.
+
+### Stream Send Buffers
+
+Frames are passed to the packetiser on a per stream basis.
+The frames must be fully formed.  By passing a frame to this function,
+ownership is passed to the packetiser which queues the frames for later
+sending by the record layer.
+
+```c
+__owur int ossl_quic_packetise(SSL *stream, const OSSL_QUIC_FRAME *frame,
+                               size_t frame_length);
+```
```

Comment:
This is using a stream SSL object.  I don't think the packetiser needs to maintain much, if any, per stream data but it definitely needs access to quite a bit of per stream information.  It also needs to jump up to the parent connection for a multitude of things which seems easiest to do with an SSL object -- I guess this could use a packetiser structure.  All the hooks to the other parts of the design will also be stored in the SSL objects, so I figured it was easiest to grab them from there instead of passing lots of pointers through the APIs or structures.

A bit less modular perhaps, but excessive modularity is a curse for code comprehension :grinning:

Reply:
If we go with the #18612 I do not think we should pass the polymorphic SSL * object around internally, instead, we should use the QUIC_CONNECTION * or QUIC_STREAM * as appropriate. Note that due to MVP requirements we do not have QUIC_STREAM * yet. Perhaps we might just alias it to QUIC_CONNECTION for now?